File: wpilibj/src/test/java/edu/wpi/first/wpilibj/kinematics/SwerveDriveKinematicsTest.java
/*----------------------------------------------------------------------------*/
/* Copyright (c) 2019 FIRST. All Rights Reserved.                             */
/* Open Source Software - may be modified and shared by FRC teams. The code   */
/* must be accompanied by the FIRST BSD license file in the root directory of */
/* the project.                                                               */
/*----------------------------------------------------------------------------*/

package edu.wpi.first.wpilibj.kinematics;

import org.junit.jupiter.api.Test;

import edu.wpi.first.wpilibj.geometry.Rotation2d;
import edu.wpi.first.wpilibj.geometry.Translation2d;

import static org.junit.jupiter.api.Assertions.assertAll;
import static org.junit.jupiter.api.Assertions.assertEquals;

class SwerveDriveKinematicsTest {
  private static final double kEpsilon = 1E-9;

  private final Translation2d m_fl = new Translation2d(12, 12);
  private final Translation2d m_fr = new Translation2d(12, -12);
  private final Translation2d m_bl = new Translation2d(-12, 12);
  private final Translation2d m_br = new Translation2d(-12, -12);

  private final SwerveDriveKinematics m_kinematics =
      new SwerveDriveKinematics(m_fl, m_fr, m_bl, m_br);

  @Test
  void testStraightLineInverseKinematics() {

    ChassisSpeeds speeds = new ChassisSpeeds(5, 0, 0);
    var moduleStates = m_kinematics.toSwerveModuleStates(speeds);

    assertAll(
        () -> assertEquals(5.0, moduleStates[0].speed, kEpsilon),
        () -> assertEquals(5.0, moduleStates[1].speed, kEpsilon),
        () -> assertEquals(5.0, moduleStates[2].speed, kEpsilon),
        () -> assertEquals(5.0, moduleStates[3].speed, kEpsilon),
        () -> assertEquals(0.0, moduleStates[0].angle.getRadians(), kEpsilon),
        () -> assertEquals(0.0, moduleStates[1].angle.getRadians(), kEpsilon),
        () -> assertEquals(0.0, moduleStates[2].angle.getRadians(), kEpsilon),
        () -> assertEquals(0.0, moduleStates[3].angle.getRadians(), kEpsilon)
    );
  }

  @Test
  void testStraightLineForwardKinematics() {
    SwerveModuleState state = new SwerveModuleState(5.0, Rotation2d.fromDegrees(90.0));
    var chassisSpeeds = m_kinematics.toChassisSpeeds(state, state, state, state);

    assertAll(
        () -> assertEquals(0.0, chassisSpeeds.vx, kEpsilon),
        () -> assertEquals(5.0, chassisSpeeds.vy, kEpsilon),
        () -> assertEquals(0.0, chassisSpeeds.omega, kEpsilon)
    );
  }

  @Test
  void testStraightStrafeInverseKinematics() {

    ChassisSpeeds speeds = new ChassisSpeeds(0, 5, 0);
    var moduleStates = m_kinematics.toSwerveModuleStates(speeds);

    assertAll(
            () -> assertEquals(5.0, moduleStates[0].speed, kEpsilon),
            () -> assertEquals(5.0, moduleStates[1].speed, kEpsilon),
            () -> assertEquals(5.0, moduleStates[2].speed, kEpsilon),
            () -> assertEquals(5.0, moduleStates[3].speed, kEpsilon),
            () -> assertEquals(90.0, moduleStates[0].angle.getDegrees(), kEpsilon),
            () -> assertEquals(90.0, moduleStates[1].angle.getDegrees(), kEpsilon),
            () -> assertEquals(90.0, moduleStates[2].angle.getDegrees(), kEpsilon),
            () -> assertEquals(90.0, moduleStates[3].angle.getDegrees(), kEpsilon)
    );
  }

    @Test
    void testStraightStrafeForwardKinematics() {
        SwerveModuleState state = new SwerveModuleState(5.0, Rotation2d.fromDegrees(0.0));
        var chassisSpeeds = m_kinematics.toChassisSpeeds(state, state, state, state);

        assertAll(
                () -> assertEquals(5.0, chassisSpeeds.vx, kEpsilon),
                () -> assertEquals(0.0, chassisSpeeds.vy, kEpsilon),
                () -> assertEquals(0.0, chassisSpeeds.omega, kEpsilon)
        );
    }

    @Test
    void testTurnInPlaceInverseKinematics() {

        ChassisSpeeds speeds = new ChassisSpeeds(0, 0, 2*Math.PI);
        var moduleStates = m_kinematics.toSwerveModuleStates(speeds);

        /*
        The circumference of the wheels about the COR is pi * diameter, or 2 * pi * radius
        the radius is the sqrt(12^2in + 12^2in), or 16.9706in, so the circumference the wheels
        trace out is 106.629190516in. since we want our robot to rotate at 1 rotation per second,
        our wheels must trace out 1 rotation (or 106.63 inches) per second.
         */

        assertAll(
                () -> assertEquals(106.63, moduleStates[0].speed, 0.1),
                () -> assertEquals(106.63, moduleStates[1].speed, 0.1),
                () -> assertEquals(106.63, moduleStates[2].speed, 0.1),
                () -> assertEquals(106.63, moduleStates[3].speed, 0.1),
                () -> assertEquals(135.0, moduleStates[0].angle.getDegrees(), kEpsilon),
                () -> assertEquals(45.0, moduleStates[1].angle.getDegrees(), kEpsilon),
                () -> assertEquals(-135.0, moduleStates[2].angle.getDegrees(), kEpsilon),
                () -> assertEquals(-45.0, moduleStates[3].angle.getDegrees(), kEpsilon)
        );
    }

    @Test
    void testTurnInPlaceForwardKinematics() {
        SwerveModuleState fl_state = new SwerveModuleState(106.629, Rotation2d.fromDegrees(135));
        SwerveModuleState fr_state = new SwerveModuleState(106.629, Rotation2d.fromDegrees(45));
        SwerveModuleState bl_state = new SwerveModuleState(106.629, Rotation2d.fromDegrees(-135));
        SwerveModuleState br_state = new SwerveModuleState(106.629, Rotation2d.fromDegrees(-45));

        var chassisSpeeds = m_kinematics.toChassisSpeeds(fl_state, fr_state, bl_state, br_state);

        assertAll(
                () -> assertEquals(0.0, chassisSpeeds.vx, kEpsilon),
                () -> assertEquals(0.0, chassisSpeeds.vy, kEpsilon),
                () -> assertEquals(2 * Math.PI, chassisSpeeds.omega, 0.1)
        );
    }

    @Test
    void testOffCenterCORInverseKinematics() {

        ChassisSpeeds speeds = new ChassisSpeeds(0, 0, 2*Math.PI);
        var moduleStates = m_kinematics.toSwerveModuleStates(speeds, m_fl);

        /*
        This one is a bit trickier. Because we are rotating about the front-left wheel,
        it should be parked at 0 degrees and 0 speed. The front-right and back-left wheels both travel an
        arc with radius 24 (and circumference 150.796), and the back-right wheel travels an arc with
        radius sqrt(24^2 + 24^2) and circumference 213.2584. As for angles, the front-right wheel should be
        pointing straight forward, the back-left wheel should be pointing straight right, and the back-right
        wheel should be at a -45 degree angle
         */

        assertAll(
                () -> assertEquals(0.0, moduleStates[0].speed, 0.1),
                () -> assertEquals(150.796, moduleStates[1].speed, 0.1),
                () -> assertEquals(150.796, moduleStates[2].speed, 0.1),
                () -> assertEquals(213.258, moduleStates[3].speed, 0.1),
                () -> assertEquals(0.0, moduleStates[0].angle.getDegrees(), kEpsilon),
                () -> assertEquals(0.0, moduleStates[1].angle.getDegrees(), kEpsilon),
                () -> assertEquals(-90.0, moduleStates[2].angle.getDegrees(), kEpsilon),
                () -> assertEquals(-45.0, moduleStates[3].angle.getDegrees(), kEpsilon)
        );
    }

    @Test
    void testOffCenterCORForwardKinematics() {
        SwerveModuleState fl_state = new SwerveModuleState(0.0, Rotation2d.fromDegrees(0.0));
        SwerveModuleState fr_state = new SwerveModuleState(150.796, Rotation2d.fromDegrees(0.0));
        SwerveModuleState bl_state = new SwerveModuleState(150.796, Rotation2d.fromDegrees(-90));
        SwerveModuleState br_state = new SwerveModuleState(213.258, Rotation2d.fromDegrees(-45));

        var chassisSpeeds = m_kinematics.toChassisSpeeds(fl_state, fr_state, bl_state, br_state);

        /*
        We already know that our omega should be 2pi from the previous test. Next, we need to determine
        the vx and vy of our chassis center. Because our COR is at a 45 degree angle from the center,
        we know that vx and vy must be the same. Furthermore, we know that the center of mass makes
        a full revolution about the center of revolution once every second. Therefore, the center of
        mass must be moving at 106.629in/sec. Recalling that the ratios of a 45/45/90 triagle are
        1:sqrt(2)/2:sqrt(2)/2, we find that the COM vx is -75.398, and vy is 75.398.
        */

        assertAll(
                () -> assertEquals(75.398, chassisSpeeds.vx, 0.1),
                () -> assertEquals(-75.398, chassisSpeeds.vy, 0.1),
                () -> assertEquals(2 * Math.PI, chassisSpeeds.omega, 0.1)
        );
    }

}

File: wpilibc/src/test/native/cpp/kinematics/SwerveDriveKinematicsTest.cpp
/*----------------------------------------------------------------------------*/
/* Copyright (c) 2019 FIRST. All Rights Reserved.                             */
/* Open Source Software - may be modified and shared by FRC teams. The code   */
/* must be accompanied by the FIRST BSD license file in the root directory of */
/* the project.                                                               */
/*----------------------------------------------------------------------------*/

#include <units/units.h>

#include "frc/kinematics/SwerveDriveKinematics.h"
#include "gtest/gtest.h"

using namespace frc;

static constexpr double kEpsilon = 1E-9;

TEST(SwerveDriveKinematics, StraightLineInverseKinematics) {
  ChassisSpeeds speeds{5.0_mps, 0.0_mps, 0.0_rad_per_s};

  SwerveDriveKinematics kinematics{
      Translation2d{12_m, 12_m}, Translation2d{12_m, -12_m},
      Translation2d{-12_m, 12_m}, Translation2d{-12_m, -12_m}};

  auto [fl, fr, bl, br] = kinematics.ToSwerveModuleStates(speeds);

  EXPECT_NEAR(fl.speed.to<double>(), 5.0, kEpsilon);
  EXPECT_NEAR(fr.speed.to<double>(), 5.0, kEpsilon);
  EXPECT_NEAR(bl.speed.to<double>(), 5.0, kEpsilon);
  EXPECT_NEAR(br.speed.to<double>(), 5.0, kEpsilon);

  EXPECT_NEAR(fl.angle.Radians().to<double>(), 0.0, kEpsilon);
  EXPECT_NEAR(fr.angle.Radians().to<double>(), 0.0, kEpsilon);
  EXPECT_NEAR(bl.angle.Radians().to<double>(), 0.0, kEpsilon);
  EXPECT_NEAR(br.angle.Radians().to<double>(), 0.0, kEpsilon);
}

TEST(SwerveDriveKinematics, StraightLineForwardKinematics) {
  SwerveDriveKinematics kinematics{
      Translation2d{12_m, 12_m}, Translation2d{12_m, -12_m},
      Translation2d{-12_m, 12_m}, Translation2d{-12_m, -12_m}};

  SwerveModuleState state{5.0_mps, Rotation2d()};

  auto chassisSpeeds = kinematics.ToChassisSpeeds(state, state, state, state);

  EXPECT_NEAR(chassisSpeeds.vx.to<double>(), 5.0, kEpsilon);
  EXPECT_NEAR(chassisSpeeds.vy.to<double>(), 0.0, kEpsilon);
  EXPECT_NEAR(chassisSpeeds.omega.to<double>(), 0.0, kEpsilon);
}

TEST(SwerveDriveKinematics, NormalizeTest) {
  SwerveModuleState state1{5.0_mps, Rotation2d()};
  SwerveModuleState state2{6.0_mps, Rotation2d()};
  SwerveModuleState state3{4.0_mps, Rotation2d()};
  SwerveModuleState state4{7.0_mps, Rotation2d()};

  std::array<SwerveModuleState, 4> arr{state1, state2, state3, state4};
  SwerveDriveKinematics<4>::NormalizeWheelSpeeds(&arr, 5.5_mps);

  double kFactor = 5.5 / 7.0;

  EXPECT_NEAR(arr[0].speed.to<double>(), 5.0 * kFactor, kEpsilon);
  EXPECT_NEAR(arr[1].speed.to<double>(), 6.0 * kFactor, kEpsilon);
  EXPECT_NEAR(arr[2].speed.to<double>(), 4.0 * kFactor, kEpsilon);
  EXPECT_NEAR(arr[3].speed.to<double>(), 7.0 * kFactor, kEpsilon);
}
